In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.neighbors import KNeighborsClassifier
from sklearn import preprocessing
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.linear_model import LogisticRegression
from sklearn.linear_model import LogisticRegressionCV
from sklearn.tree import DecisionTreeClassifier

df = pd.read_csv("wdbc.csv")

In [2]:
atributos = ["Diag", "Perimeter", "Area", "Compact", "Simmetry", "Texture", "AreaW", "SmoothW"]

df_clean = df.replace({"Diag":{'B':0, 'M':1}})

atributos = list(df_clean.columns.values)
atributos.remove('Diag')

X = df_clean.loc[:, atributos]
Y = df_clean.loc[:, 'Diag']

X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.25, random_state=0)

In [3]:
for i in range (1,10):
    knn = KNeighborsClassifier(n_neighbors=i)
    knn_scores = cross_val_score(knn, X_train, Y_train, cv=10)
    print(i,"-nearest-neighbors score:", np.mean(knn_scores))

1 -nearest-neighbors score: 0.8192474948005295
2 -nearest-neighbors score: 0.7721213840045377
3 -nearest-neighbors score: 0.7628771304324339
4 -nearest-neighbors score: 0.7559557572319908
5 -nearest-neighbors score: 0.7397901304594441
6 -nearest-neighbors score: 0.7395105745077385
7 -nearest-neighbors score: 0.7347459687221457
8 -nearest-neighbors score: 0.7420630418928774
9 -nearest-neighbors score: 0.7234610917537747


O algoritmo com o melhor score foi o 1-vizinhos mais próximos.

In [4]:
max_score = 0;
best_depth = 0;
best_min_sample_leaf = 0;

for i in range (1,10):
    for j in range (1,10):
        #print("Max Depth = ", i,"\nMin Samples Leaf =", j)
        dt = DecisionTreeClassifier(criterion = 'gini', max_depth = i, min_samples_leaf = j, random_state = 0)

        dt_score = (cross_val_score(dt, X_train, Y_train, cv=10))
        #print("Decision Tree score:", np.mean(dt_score))
        if(np.mean(dt_score) > max_score):
            max_score = np.mean(dt_score)
            best_depth = i;
            best_min_sample_leaf = j;

In [5]:
print("Score: ",max_score,"\nbest_depth: ",best_depth,"\nbest_min_sample_leaf: ",best_min_sample_leaf)

Score:  0.943568862598925 
best_depth:  4 
best_min_sample_leaf:  2


O melhor desempenho foi 0.93865 usando os parâmetros 4 para a profundidade e 6 para quantidade mpinima de amostras de nós folha.

In [6]:
knn = KNeighborsClassifier(n_neighbors=1)
knn_scores = cross_val_score(knn, X_test, Y_test, cv=10)
print("\n1-nearest-neighbors score:",np.mean(knn_scores))


1-nearest-neighbors score: 0.720952380952381


In [7]:
dt = DecisionTreeClassifier(criterion = 'gini', max_depth = 4, min_samples_leaf = 2, random_state = 0)
dt_score = (cross_val_score(dt, X_test, Y_test, cv=10))
print("Decision Tree score:", np.mean(dt_score))

Decision Tree score: 0.9019047619047619


A acurácia do knn com 1 vizinho foi de 82% no treino e 72% no teste, já na árvore de decisão com parâmetros 4 (profundidade) e 2 (mínimo no nó folha), a acurácia foi de 94% no treino e 90% no teste.